modules/hostfrequencyscaling/hostfrequencyscaling.go

```diff
@@ -0,0 +1,516 @@
+/* hostfrequencyscaling.go: performs mutations related to scaling of CPU frequency to control CPU thermal conditions of HPC node using in-band mechanism.
+ *
+ * This module boots HPC nodes with "schedutil" scaling governor and whenever CPU temperature reaches to high (warning) condition, module mutates the scaling governor to "powersave".
+ * Current implementation handles critical CPU temperature same as high CPU temperature.
+ *
+ * Additionally, there are many other mutations intended for different use cases (e.g. switching back to "schedutil" after "powersave") are under considerations and investinations.
+ *
+ * Authors: Ghazanfar Ali, ghazanfar.ali@ttu.edu; Kevin Pelzel <kevinpelzel22@gmail.com>; J. Lowell Wofford <lowell@lanl.gov>
+ *
+ * This software is open source software available under the BSD-3 license.
+ * Copyright (c) 2019, Triad National Security, LLC
+ * See LICENSE file for details.
+ */
+
+package hostfrequencyscaling
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/hpc/kraken/core"
+	cpb "github.com/hpc/kraken/core/proto"
+	hostthpb "github.com/hpc/kraken/extensions/HostThermal/proto"
+	rp3pb "github.com/hpc/kraken/extensions/RPi3/proto"
+	"github.com/hpc/kraken/lib"
+	pb "github.com/hpc/kraken/modules/hostfrequencyscaling/proto"
+)
+
+// CPUPerfScalingReq is payload for RFAggregator API call
+type CPUPerfScalingReq struct {
+	ScalingGovernor  string   `json:"scalinggovernor"`
+	ScalingMinFreq   string   `json:"scalingminfreq"`
+	ScalingMaxFreq   string   `json:"scalingmaxfreq"`
+	NodesAddressList []string `json:"nodesaddresslist,omitempty"`
+	Timeout          int      `json:"timeout,omitempty"`
+}
+
+// CPUPerfScalingResp structure for API response from RFAggregator
+type CPUPerfScalingResp struct {
+	TimeStamp          time.Time `json:"timestamp"`
+	HostAddress        string    `json:"hostaddress"`
+	CurScalingGovernor string    `json:"curscalinggovernor"`
+	ScalingMinFreq     string    `json:"scalingminfreq"`
+	ScalingMaxFreq     string    `json:"scalingmaxfreq"`
+	ScalingCurFreq     string    `json:"scalingcurfreq"`
+
+	CPUCurFreq string `json:"cpucurfreq"`
+	CPUMinFreq string `json:"cpuminfreq"`
+	CPUMaxFreq string `json:"cpumaxfreq"`
+}
+
+// CPUPerfScalingRespColl structure for collection of response
+type CPUPerfScalingRespColl struct {
+	CPUPerfScalingRespCollection []CPUPerfScalingResp `json:"cpuperfscalingrespcollection"`
+}
+
+const (
+	// PxeURL refers to PXE object
+	PxeURL string = "type.googleapis.com/proto.RPi3/Pxe"
+
+	// ModuleStateURL refers to module state
+	ModuleStateURL string = "/Services/hostfrequencyscaling/State"
+
+	// HostThermalStateURL points to Thermal extension
+	HostThermalStateURL string = "type.googleapis.com/proto.HostThermal/State"
+
+	// NodeIPURL provides node IP address
+	NodeIPURL string = "type.googleapis.com/proto.IPv4OverEthernet/Ifaces/0/Ip/Ip"
```

Comment:
These constant values _probably_ don't need to be exported, but it probably doesn't really matter. Ask Lowell